src/realm/deppart/partitions_gpu_impl.hpp

```diff
@@ -0,0 +1,343 @@
+    }                                                                        \
+  } while (0)
+
+
+//NVTX macros to only add ranges if defined
+#ifdef REALM_USE_NVTX
+
+  #define NVTX_CAT(a,b)  a##b
+
+  #define NVTX_DEPPART(message) \
+  nvtxScopedRange NVTX_CAT(nvtx_, message)("cuda", #message, 0)
+
+#else
+
+  #define NVTX_DEPPART(message) do { } while (0)
+
+#endif
+
+namespace Realm {
+
+  //Used by cub::DeviceReduce to compute bad GPU approximation
+  template<int N, typename T>
+  struct UnionRectOp {
+    __host__ __device__
+    Rect<N,T> operator()(const Rect<N,T>& a,
+                         const Rect<N,T>& b) const {
+      Rect<N,T> r;
+      for(int d=0; d<N; d++){
+        r.lo[d] = a.lo[d] <  b.lo[d] ? a.lo[d] : b.lo[d];
+        r.hi[d] = a.hi[d] > b.hi[d] ? a.hi[d] : b.hi[d];
+      }
+      return r;
+    }
+  };
+
+
+  /*
+   *  Input: An array of points (potentially with duplicates) with associated
+   *  src indices, where all the points with a given src idx together represent an exact covering
+   *  of the partitioning output for that index.
+   *  Output: A disjoint, coalesced array of rectangles sorted by src idx that it then sends off
+   *  to the send output function, which constructs the final sparsity map.
+   *  Approach: Sort the points by (x0,x1,...,xN-1,src) (right is MSB). Convert them to singleton rects.
+   *  Run-length encode along each dimension (N-1...0).
+   */
+  template<int N, typename T>
+  template<typename Container, typename IndexFn, typename MapFn>
+  void GPUMicroOp<N,T>::complete_pipeline(PointDesc<N, T>* d_points, size_t total_pts, Memory my_mem, const Container& ctr, IndexFn getIndex, MapFn getMap)
+  {
+
+    NVTX_DEPPART(complete_pipeline);
+
+    cudaStream_t stream = Cuda::get_task_cuda_stream();
+
+    size_t bytes_T   = total_pts * sizeof(T);
+    size_t bytes_S   = total_pts * sizeof(size_t);
+    size_t bytes_R  =  total_pts * sizeof(RectDesc<N,T>);
+    size_t bytes_p = total_pts * sizeof(PointDesc<N,T>);
+    size_t max_aux_bytes = std::max({bytes_T, bytes_S, bytes_R});
+    size_t max_pg_bytes = std::max({bytes_p, bytes_S});
+
+
+    // Instance shared by coordinate keys, source keys, and rectangle outputs
+    RegionInstance aux_instance = this->realm_malloc(2 * max_aux_bytes, my_mem);
+
+    //Instance shared by group ids (RLE) and intermediate points in sorting
+    RegionInstance pg_instance = this->realm_malloc(max_pg_bytes, my_mem);
+
+    RegionInstance break_points_instance = this->realm_malloc(total_pts * sizeof(uint8_t), my_mem);
+
+    const uintptr_t aux_in = AffineAccessor<char,1>(aux_instance, 0).base;
+    const uintptr_t aux_out = aux_in + max_aux_bytes;
+
+    T* d_keys_in = reinterpret_cast<T*>(aux_in);
+    T* d_keys_out = reinterpret_cast<T*>(aux_out);
+
+    PointDesc<N,T>* d_points_in = d_points;
+    PointDesc<N,T>* d_points_out = reinterpret_cast<PointDesc<N,T>*>(AffineAccessor<char,1>(pg_instance, 0).base);
```

Comment:
Why are you doing reinterpret cases instead of making the affine accessor with the right type?